design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+name: example-appconfig
+spec:
+components:
+ - componentName: example-db      
+   traits:
+     - trait:
+         apiVersion: core.oam.dev/v1alpha2
+         kind: ManualScalerTrait
+         metadata:
+           name:  example-appconfig-trait
+         spec:
+           replicaCount: 3
+```
+
+The problem is two folds
+1. A trait controller needs a way to find the workload that it is applied to. 
+   - In the example, the manualScalar trait needs to know that it is supposed to scale the
+    example-db workload. However, since the applicationConfiguration controller just emit a
+     `ManualScalerTrait` CR, it does not contain any information on the `example-db` workload.
+2. A trait controller needs to know the exact resources it should modify. Note that these
+resources are most likely not the workload itself.
+    - Use the same example, just knowing the `example-db` workload is not enough for the
+     ManualScalerTrait to work. The trait controller does not work with the `example-db` workload
+      directly. It needs to find the actual Kubernetes resources that the `example-db`
+     workload generates and then it can modify the `replica` field in its spec.
+
+## Goals
+In order to maximize the extensibility of our OAM implementation, our solution need to meet the
+ following two design objections.
+1. **Extensible trait system**: We want to allow a `trait` to apply to any eligible `workload`
+instead of just a list of specific ones. This means that we want to empower a trait developer to
+write the controller code once and it will work for any new `workload` that this `trait` can
+apply to in the future.
+    - Using the example again, the `ManualScalerTrait` should work with any workload that
+     generates a Kubernetes resource that has a `replica` field in its spec even if the
+      workload does not exist when the `ManualScalerTrait` is implemented.
+2. **Adopting existing CRDs**: The mechanism cannot put any limit on the `trait` or `workload
+` CRDs. This means that we cannot assume any pre-defined CRD fields in any `trait` or `workload
+` beyond Kubernetes conventions (i.e. spec or status).
+    - While an additional field in the `ManualScalerTrait` spec that points to a workload could
+     partially alleviate the problem, this is not the type of solution we are aiming for. This is
+     because OAM platform builders or infrastructure teams can install any existing CRDs they
+     maintained to the platform as a `workload` or `trait`. Therefore, our solution needs to
+     accommodate pre-defined CRDs that OAM users bring.
+
+## Proposal
+The overall idea is for the applicationConfiguration controller to annotate critical information
+in the workload and trait CR it emits. In addition, we will provide a helper library so that
+trait controller developers can locate the resources they need with a simple function call. 
+Here is the list of changes that we propose.
+1. Add a `childResources` field in the  WorkloadDefinition. 
```

Comment:
My immediate thought here was that it feels a little weird to be defining on the `WorkloadDefinition` something that is enforced directly by code of the workload controller. However, I actually think this is a fine pattern as the `WorkloadDefinition` will likely be installed by the system that is implementing the workload controllers, similar to how a Kubernetes operator installs its CRDs 👍